flink-connectors/flink-connector-elasticsearch7/src/main/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchWriter.java

```diff
@@ -0,0 +1,326 @@
+     *
+     * @param hosts the reachable elasticsearch cluster nodes
+     * @param processor converting incoming records to elasticsearch actions
+     * @param flushOnCheckpoint if true all until now received records are flushed after every
+     *     checkpoint
+     * @param bulkProcessorConfig describing the flushing and failure handling of the used {@link
+     *     BulkProcessor}
+     * @param networkClientConfig describing properties of the network connection used to connect to
+     *     the elasticsearch cluster
+     * @param metricGroup for the sink writer
+     * @param mailboxExecutor Flink's mailbox executor
+     */
+    public ElasticsearchWriter(
+            List<HttpHost> hosts,
+            ElasticsearchProcessor<IN> processor,
+            boolean flushOnCheckpoint,
+            BulkProcessorConfig bulkProcessorConfig,
+            NetworkClientConfig networkClientConfig,
+            SinkWriterMetricGroup metricGroup,
+            MailboxExecutor mailboxExecutor) {
+        this.processor = checkNotNull(processor);
+        this.flushOnCheckpoint = flushOnCheckpoint;
+        this.mailboxExecutor = checkNotNull(mailboxExecutor);
+        this.client =
+                new RestHighLevelClient(
+                        configureRestClientBuilder(
+                                RestClient.builder(hosts.toArray(new HttpHost[0])),
+                                networkClientConfig));
+        this.bulkProcessor =
+                configureBulkProcessor(
+                        BulkProcessor.builder(
+                                (bulkRequest, bulkResponseActionListener) ->
+                                        client.bulkAsync(
+                                                bulkRequest,
+                                                RequestOptions.DEFAULT,
+                                                bulkResponseActionListener),
+                                new BulkListener()),
+                        bulkProcessorConfig);
+        this.requestIndexer = new DefaultRequestIndexer();
+        checkNotNull(metricGroup);
+        metricGroup.setCurrentSendTimeGauge(() -> receiveTime - lastSendTime);
+        this.numBytesOutCounter = metricGroup.getIOMetricGroup().getNumBytesOutCounter();
+        try {
+            processor.open();
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to open the ElasticsearchProcessor", e);
+        }
+    }
+
+    @Override
+    public void write(IN element, Context context) throws IOException, InterruptedException {
+        // do not allow new bulk writes until all actions are flushed
+        while (checkpointInProgress) {
+            mailboxExecutor.yield();
+        }
+        processor.process(element, context, requestIndexer);
+    }
+
+    @Override
+    public List<Void> prepareCommit(boolean flush) throws IOException, InterruptedException {
+        checkpointInProgress = true;
+        while (pendingActions != 0) {
```

Comment:
Here we should also flush for `!flushOnCheckpoint && flush`. That means we always count pending elements.